### DISCUSSION TITLE
How to render nodes and edges for static HTML generation?

### DISCUSSION THREAD

**Discussion by timdohm3f** [xyflow/xyflow#4078] [Q&A] [answered]
I'm trying to generate a Reactflow diagram to be embedded in a PDF. We generate this PDF by first generating static HTML by essentially doing this:
```
import ReactDOMServer from 'react-dom/server';
const el: React.ReactNode = getReactNode();
const finalHtml = ReactDOMServer.renderToStaticMarkup(el);
```
The Reactflow diagram renders properly in Storybook, as well as in a web page we generate. However, when doing the static HTML generation, as above, the nodes and edges don't show up. Other elements (e.g. `<Background/>`, `<Panel/>`) do render but not nodes/edges. 

Any thoughts on why this is and how this might be resolved?

**Comment by moklick** [comment]
this is only possible with React Flow 12, which is still alpha https://github.com/xyflow/xyflow/discussions/3764

**Reply by timdohm3f** [reply]
Thanks for that. I converted my project to v12 but after some effort, I could only get a few text and SVG elements to render. No custom nodes or edges showed up.

**Reply by timdohm3f** [reply]
I actually have this working fairly well now (issue was missing Handles.) My only issue now is that `fitview` doesn't work and I'm unable to see the entire graph when it's rendered in the PDF. I've tried using `getViewportForBounds` to generate a viewport but the viewport doesn't apply in the `ReactFlow` component at render time. Any suggestions on this?

**Reply by moklick** [reply, marked as answer]
For `fitView` on the server, you need to pass a `width` and `height` prop to the `ReactFlow` component. Does that fix it?

**Reply by timdohm3f** [reply]
Yes, that fixed it. Thank you so much!
